lib/utils/compose.coffee

```diff
@@ -0,0 +1,747 @@
+	.get('body')
+	.get('token')
+	.catchReturn({})
+
+pushAndUpdateServiceImages = (docker, token, images, afterEach) ->
+	chalk = require('chalk')
+	{ DockerProgress } = require('docker-progress')
+	tty = require('./tty')(process.stdout)
+
+	opts = { authconfig: registrytoken: token }
+
+	progress = new DockerProgress(dockerToolbelt: docker)
+	renderer = pushProgressRenderer(tty, chalk.blue('[Push]') + '    ')
+	reporters = progress.aggregateProgress(images.length, renderer)
+
+	Promise.using tty.cursorHidden(), ->
+		Promise.map images, ({ serviceImage, localImage, props, logs }, index) ->
+			Promise.join(
+				localImage.inspect().get('Size')
+				progress.push(localImage.name, reporters[index], opts).finally(renderer.end)
+				(size, digest) ->
+					serviceImage.image_size = size
+					serviceImage.content_hash = digest
+					serviceImage.build_log = logs
+					serviceImage.dockerfile = props.dockerfile
+					serviceImage.project_type = props.projectType
+					serviceImage.start_timestamp = props.startTime if props.startTime
+					serviceImage.end_timestamp = props.endTime if props.endTime
+					serviceImage.push_timestamp = new Date()
+					serviceImage.status = 'success'
+			)
+			.tapCatch (e) ->
+				serviceImage.error_message = '' + e
+				serviceImage.status = 'failed'
+			.finally ->
+				afterEach?(serviceImage, props)
+
+exports.deployProject = (
+	docker, logger,
+	composition, images,
+	appId, userId, auth,
+	apiEndpoint,
+	skipLogUpload
+) ->
+	_ = require('lodash')
+	chalk = require('chalk')
+	releaseMod = require('resin-release')
+	tty = require('./tty')(process.stdout)
+
+	prefix = chalk.cyan('[Info]') + '    '
+	spinner = createSpinner()
+	runloop = runSpinner(tty, spinner, "#{prefix}Creating release...")
+
+	createRelease(apiEndpoint, auth, userId, appId, composition)
+	.finally(runloop.end)
+	.then ({ client, release, serviceImages }) ->
+		logger.logDebug('Tagging images...')
+		tagServiceImages(docker, images, serviceImages)
+		.tap (images) ->
+			logger.logDebug('Authorizing push...')
+			authorizePush(apiEndpoint, auth, images[0].registry, _.map(images, 'repo'))
+			.then (token) ->
+				logger.logInfo('Pushing images to registry...')
+				pushAndUpdateServiceImages docker, token, images, (serviceImage) ->
+					logger.logDebug("Saving image #{serviceImage.is_stored_at__image_location}")
+					if skipLogUpload
+						delete serviceImage.build_log
+					releaseMod.updateImage(client, serviceImage.id, serviceImage)
+			.finally ->
+				logger.logDebug('Untagging images...')
+				Promise.map images, ({ localImage }) ->
+					localImage.remove()
+		.then ->
+			release.status = 'success'
+		.tapCatch (e) ->
+			release.status = 'failed'
+		.finally ->
+			runloop = runSpinner(tty, spinner, "#{prefix}Saving release...")
+			release.end_timestamp = new Date()
+			releaseMod.updateRelease(client, release.id, release)
+			.finally(runloop.end)
+
+# utilities
+
+renderProgressBar = (percentage, stepCount) ->
```

Comment:
Don't we already have something like this in resin-cli-visuals?

Reply:
Yeah, Tim also mentioned this here: https://github.com/resin-io/resin-cli/pull/763#discussion_r165714077